dockstore-webservice/src/main/java/io/dockstore/webservice/resources/EntryResource.java

```diff
@@ -421,6 +433,101 @@ public Entry setDiscourseTopic(@ApiParam(hidden = true) @Parameter(hidden = true
+        return progress.processedEntries;
+    }
+
+    private int updateLanguageVersionsForLegacyTools(boolean allVersions) {
+        final LanguageVersionProgress progress = new LanguageVersionProgress();
+        while (!progress.done) {
+            final TransactionHelper transactionHelper = new TransactionHelper(sesssionFactory);
+            transactionHelper.transaction(() -> {
+                final List<Tool> tools =
+                    this.toolDAO.findAllTools(progress.offset, LANGUAGE_VERSION_PROCESSOR_PAGE_SIZE);
+                if (tools.size() < LANGUAGE_VERSION_PROCESSOR_PAGE_SIZE) {
+                    progress.done = true;
+                }
+                progress.processedEntries += tools.size();
+                LOG.info("Executing {} tool language version updates starting at offset {}",
+                    LANGUAGE_VERSION_PROCESSOR_PAGE_SIZE, progress.offset);
+                progress.offset += LANGUAGE_VERSION_PROCESSOR_PAGE_SIZE;
+                tools.stream()
+                    .filter(tool -> tool.getDescriptorType().size() == 1) // Only support tools with 1 language
+                    .forEach(tool -> {
+                        final String descriptorLanguageText = tool.getDescriptorType().get(0);
+                        LanguageHandlerInterface languageHandlerInterface = LanguageHandlerFactory.getInterface(
+                            DescriptorLanguage.convertShortStringToEnum(descriptorLanguageText));
+                        tool.getWorkflowVersions().stream()
+                            .filter(version -> allVersions || version.getVersionMetadata().getDescriptorTypeVersions().isEmpty())
+                            .forEach(version -> {
+                                final String primaryPath;
+                                if (DescriptorLanguage.convertShortStringToEnum(descriptorLanguageText) == DescriptorLanguage.WDL) {
+                                    primaryPath = version.getWdlPath();
+                                } else { // We will not add new languages to Tool (perhaps to AppTool), so this is safe
+                                    primaryPath = version.getCwlPath();
+                                }
+                                readSourceFilesAndUpdate(languageHandlerInterface, version, primaryPath);
+                            });
+                    });
+            });
+        }
+        LOG.info("Completed processing {} legacy tools", progress.processedEntries);
+        return progress.processedEntries;
+    }
+
+    private void readSourceFilesAndUpdate(final LanguageHandlerInterface languageHandlerInterface,
+        final Version<? extends Version> workflowVersion, String primaryPath) {
+        workflowVersion.getSourceFiles().stream()
+            .filter(sf -> isPrimaryDescriptor(primaryPath, sf))
+            .findFirst()
+            .ifPresent(primary ->
+                languageHandlerInterface.parseWorkflowContent(primaryPath,
```

Comment:
Do we need to handle exceptions if `parseWorkflowContent` throws?

Reply:
Hmm, it doesn't declare that it throws anything, but at a glance I see that at least WDLHandler can throw a CustomWebApplicationException, so yes, good catch.